pyccel/parser/semantic.py

```diff
@@ -510,6 +513,28 @@ def _infer_type(self, expr, **settings):
             d_var['cls_base'   ] = NumpyArrayClass
             return d_var
 
+        elif isinstance(expr, CupyNewArray):
+            d_var['datatype'   ] = expr.dtype
+            d_var['memory_handling'] = 'heap' if expr.rank > 0 else 'stack'
+            d_var['memory_location'] = expr.memory_location
+            d_var['shape'      ] = expr.shape
+            d_var['rank'       ] = expr.rank
+            d_var['order'      ] = expr.order
+            d_var['precision'  ] = expr.precision
+            d_var['cls_base'   ] = CudaArrayClass
+            return d_var
+
+        elif isinstance(expr, CudaNewArray):
+            d_var['datatype'   ] = expr.dtype
+            d_var['memory_handling'] = 'heap' if expr.rank > 0 else 'stack'
+            d_var['memory_location'] = expr.memory_location
+            d_var['shape'      ] = expr.shape
+            d_var['rank'       ] = expr.rank
+            d_var['order'      ] = expr.order
+            d_var['precision'  ] = expr.precision
+            d_var['cls_base'   ] = CudaArrayClass
+            return d_var
```

Comment:
@EmilyBourne
- not sure if we should have two of these
- not sure why we have two of these

Reply:
Ah no they're not the same. It's `CudaNewArray` and `CupyNewArray`. Do we need both these classes though? I still haven't understood the difference if we do then it should be `elif isinstance(expr, (CudaNewArray, CupyNewArray))`

Reply:
yeah it should just be one of them as `Cupy` arrays should inherit `CudaNewArray`

Reply:
Ok. This just moves the question though. What is the difference between `CudaArray` and `CupyArray`?

Reply:
`CudaArray` is an internal function that uses `Ccuda` API for allocations.
it has more freedom to allocate to the `device`/`host`/`managed` memory locations.

`CupyArray` follows the Cupy and only allocate to the `device`.

Reply:
Maybe I'm just being stupid but I still don't see a need for both of these types. Isn't a `CupyArray` equivalent to a `CudaArray(memory_location='device')`?

When do you use `CudaArray` rather than `CupyArray`? Is a `CupyArray` converted to a `CudaArray` for printing? What Python code generates a `CudaArray`?

Reply:
I agree it needs cleanup.